pkg/lockfile/lockfile.go

```diff
@@ -401,9 +415,41 @@ func (l *LockFile) lock(lType lockType) {
 		// Optimization: only use the (expensive) syscall when
 		// the counter is 0.  In this case, we're either the first
 		// reader lock or a writer lock.
-		lockHandle(l.fd, lType)
+		lockHandle(l.fd, lType, false)
 	}
 	l.lockType = lType
 	l.locked = true
 	l.counter++
 }
+
+// lock locks the lockfile via syscall based on the specified type and
+// command.
+func (l *LockFile) tryLock(lType lockType) error {
+	if lType == readLock {
+		l.rwMutex.RLock()
+	} else {
+		l.rwMutex.Lock()
+	}
+	l.stateMutex.Lock()
+	defer l.stateMutex.Unlock()
+	if l.counter == 0 {
+		// If we're the first reference on the lock, we need to open the file again.
+		fd, err := openLock(l.file, l.ro)
+		if err != nil {
+			return err
+		}
+		l.fd = fd
+
+		// Optimization: only use the (expensive) syscall when
+		// the counter is 0.  In this case, we're either the first
+		// reader lock or a writer lock.
+		if err = lockHandle(l.fd, lType, true); err != nil {
+			closeHandle(fd)
```

Comment:
I've splitted the functions so that we don't call `UnlockFileEx` on Windows without having a lock.  Not sure how it is handled, as I've not tested it (nor I've a way to do it), but it seems incorrect to call unlock if we don't hold the lock.

Reply:
I asked for `unlockAndCloseHandle` in https://github.com/containers/storage/pull/1662#discussion_r1256114313 to make it clear to non-UNIX programmers (or programmers unfamiliar with this weird UNIX gotcha) that a non-unlocking close is impossible.

But then https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex says that `CloseHandle` also unlocks locks…

---

WDYT about `unlockAndCloseHandle(explicitClose bool)`?

Ultimately I can live with the two functions split, as the PR is now. Maybe add a comment warning about the automatic unlock?